### PR TITLE
Improve documentation for presets and outputs

### PR DESCRIPTION
**Pull request by MariusSchiffer**
Presets were not included yet in the documentation

### CODE REVIEW

File: mkosi.md [hunk 1]
```diff
@@ -1322,6 +1328,26 @@ turn-around times for complete image builds are minimal, as only
 changed source files need to be recompiled: an OS image rebuilt will
 be almost as quick to build the source tree only.
 
+# PRESETS
+
+Presets can be defined in mkosi.presets/. A preset is just like a
```
Comment (keszybz) [inline mkosi.md:1333]: Backticks around the path, here and below.

I think this text could be a bit more clear.
1. what is a preset?
2. where are presets located and how are they read?
3. how are presets ordered?
4. how are presets executed?
5. comments and examples

**Review by keszybz — CHANGES_REQUESTED** [review]

**Comment by keszybz** [comment]
@MariusSchiffer ?

**Comment by MariusSchiffer** [comment]
Ah sorry, I got sidetracked. I have a new draft, will push it soon.

File: mkosi.md [hunk 2]
```diff
@@ -151,6 +151,12 @@ Those settings cannot be configured in the configuration files.
   option is an effective way to build a project located in a specific
   directory.
 
+`--preset=`
+
+: Run only the specified preset from the `mkosi.presets` directory.
+  Can be specified multiple times to process all specified presets.
+  Regardless, presets are always processed alphabetically.
```
Comment (keszybz) [inline mkosi.md:158]: Regardless of what? It's not clear to me when I read this.

File: mkosi.md [hunk 3]
```diff
@@ -1329,6 +1335,70 @@ turn-around times for complete image builds are minimal, as only
 changed source files need to be recompiled: an OS image rebuilt will
 be almost as quick to build the source tree only.
 
+# PRESETS
+
+A preset is just like a regular configuration, except that mkosi can
```
Comment (keszybz) [inline mkosi.md:1340]: Either "like a regular configuration file" or "like regular configuration" (configuration is uncountable).

File: mkosi.md [hunk 4]
```diff
@@ -1329,6 +1335,70 @@ turn-around times for complete image builds are minimal, as only
...
+build multiple presets sequentially. Presets can be used to separate builds into stages,
+to build InitRds, or to build tools used at a later point.
```
Comment (keszybz) [inline mkosi.md:1342]: Just "initrds" please.

File: mkosi.md [hunk 5]
```diff
@@ -1329,6 +1335,70 @@ turn-around times for complete image builds are minimal, as only
...
+
+Presets are located as subdirectories in **`mkosi.presets/`**.
```
Comment (keszybz) [inline mkosi.md:1344]: Maybe "Presets are loaded from …"

File: mkosi.md [hunk 6]
```diff
@@ -1329,6 +1335,70 @@ turn-around times for complete image builds are minimal, as only
...
+If this path exists, for each preset mkosi will read the global configuration, followed by
```
Comment (keszybz) [inline mkosi.md:1345]: Strike "If this path exists, ".

File: mkosi.md [hunk 7]
```diff
@@ -1329,6 +1335,70 @@ turn-around times for complete image builds are minimal, as only
...
+the individual preset configuration. 
+
+mkosi reads and builds each of the presets in alpha-numerical order. Ordering can be enforced with a
+numeric prefix
```
Comment (keszybz) [inline mkosi.md:1349]: Period missing. But I think it'd be better to phrase this more like a hint: "Numerical prefixes can be used to make the ordering clear."

File: mkosi.md [hunk 8]
```diff
@@ -1329,6 +1335,70 @@ turn-around times for complete image builds are minimal, as only
...
+Alternatively, the presets to build can be specified as command line arguments with `--preset`.
+
+Later presets can use outputs of earlier presets.
+Output files are named based on the preset name, although without any numeric prefix.
```
Comment (keszybz) [inline mkosi.md:1353]: "numerical"

Hmm, so I don't know how this stripping works, but this explanation is not enough to figure this out. Please be explicit: what prefix is removed.

File: mkosi.md [hunk 9]
```diff
@@ -1329,6 +1335,70 @@ turn-around times for complete image builds are minimal, as only
...
+Paths in the `mkosi.conf` are relative to itself.
```
Comment (keszybz) [inline mkosi.md:1354]: "to it" ?

File: mkosi.md [hunk 10]
```diff
@@ -1329,6 +1335,70 @@ turn-around times for complete image builds are minimal, as only
...
+For example, a base preset with `Output=directory` produces a `base` directory in
+the output directory. This can be used in a later preset with `BaseTrees=base`, where
+the `base` output is copied in at the start.
```
Comment (keszybz) [inline mkosi.md:1357]: Hmm, this is not clear at all. Please put markup around all words that are actually names (e.g. `base`?).

File: mkosi.md [hunk 11]
```diff
@@ -1329,6 +1335,70 @@ turn-around times for complete image builds are minimal, as only
...
+In conjunction with the `Overlay=` option, a system extension image can be generated.
+Presets are also useful to build custom initrds.
+They can be generated with `Output=cpio` and used for UKI generation in a later
+preset with the `Initrds=` option.
+
+## Examples
+
+This example build a base filesystem that is used in a InitRd and the root filesystem.
```
Comment (keszybz) [inline mkosi.md:1365]: "builds", "an initrd"

File: mkosi.md [hunk 12]
```diff
@@ -1329,6 +1335,70 @@ turn-around times for complete image builds are minimal, as only
...
+The base filesystem could e.g. contain systemd components that are used in both the InitRd and the root filesystem.
+
+Each preset directory can additionally contain scripts that builds the respective preset.
+A more in-depth example can be found in the main systemd repository: https://github.com/systemd/systemd/tree/main/mkosi.presets
```
Comment (keszybz) [inline mkosi.md:1369]: This paragraph should be after the example.

File: mkosi.md [hunk 13]
```diff
@@ -1329,6 +1335,70 @@ turn-around times for complete image builds are minimal, as only
...
+
+
+```
+# mkosi.presets/00-base/mkosi.conf
+[Output]
+Format=directory
+
+[Content]
+Packages=build-essential
```
Comment (keszybz) [inline mkosi.md:1378]: I think the example needs to specify that this is Debian. Otherwise `build-essential` has no meaning.

File: mkosi.md [hunk 14]
```diff
@@ -1344,6 +1414,10 @@ variables:
   support out-of-tree builds to reuse already built artifacts from
   previous runs.
 
+* `$OUTPUTDIR` is set to mkosi's output directory. This can be used
+  to produce custom artifacts, for example in conjunction
+  with `Output=none`.
```
Comment (keszybz) [inline mkosi.md:1419]: Hmm, the sentence about `Output=none` is completely out of the blue.

**Review by keszybz — CHANGES_REQUESTED** [review]
This is useful, but it needs to read more like a reference document. The have to assume that the reader doesn't know too much about mkosi, is looking up just a specific option or concept. The description should be self-contained; we cannot assume that the document is read from the top to bottom. Also, the description needs to be more complete, so that the user can figure out what will happen, and not just the *intent* of the option.

**Pull request closed without merging**